designs/generics-design.md

```diff
@@ -0,0 +1,403 @@
+## Test Cases
+---
+
+Generics are a programming language feature that allow classes and functions to
+abstract over the specific types of data that they work with. This is done by
+parameterizing classes and functions using type variables and type parameters.
+
+1. **Basic Type Variable Declaration**
+
+```python
+T = TypeVar('T')
+```
+
+A type variable can be introduced using a TypeVar definition. New type variables
+can be introduced only in global scope and can be used across any number of generic
+class/function definitions.
+
+The first parameter to TypeVar can be used to provide a more descriptive name
+for the type variable and can potentially be used for better error
+messages from the compiler for example.
+
+2. **Class Definition with a Type Variable**
+
+```python
+T = TypeVar('T')
+
+class Box(Generic[T]):
+    pass
+```
+
+A generic class can be defined by inheriting from the special base class Generic
+along with a defined type variable. Any other class that this class
+inherits from in the inheritance hierarchy can be specified after this.
+
+A generic class is type-checked for all possible allowed instantiations of its
+type variables independent of the instantiations actually made in the program.
+
+3. **Object creation using Type Variable instantiated with a Primitive Type**
+```python
+T = TypeVar('T')
+
+class Box(Generic[T]):
+    pass
+
+b: Box[int] = None
+b = Box()
+```
+
+An object of a generic class can be created by instantiating the type parameter
+with a primitive type in the type annotation for the object's variable declaration.
+
+After type-checking and before lowering a copy of each generic class is made for
+each type that its type variables are instantiated with over the course of the program.
+
+
+4. **Object creation using Type Variable instantiated with a User-Defined Type**
+
+```python
+T = TypeVar('T')
+
+class Box(Generic[T]):
+    pass
+
+class Rat():
+    pass
+
+b: Box[Rat] = None
+b = Box()
+```
+
+A type parameter of a generic class can also be instantiated with a user-defined
+class.
+
+5. **Type Variables and Generic Classes with Constraints**
+
+```python
+T = TypeVar('T')
+R = TypeVar('R', int, bool)
```

Comment:
I think we could leave constraints for a future milestone – there's a lot to do just for the unconstrained version.